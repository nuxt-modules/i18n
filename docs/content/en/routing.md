---
title: Routing
description: "**@nuxtjs/i18n** overrides Nuxt default routes to add locale prefixes to every URL (except in no_prefix strategy)."
position: 6
category: Guide
---

**@nuxtjs/i18n** overrides Nuxt default routes to add locale prefixes to every URL (except in `no_prefix` strategy).
Say your app supports two languages: French and English as the default language, and you have the following pages in your project:

```asciidoc
pages/
├── index.vue
├── about.vue
```

This would result in the following routes being generated

```js
[
  {
    path: "/",
    component: _3237362a,
    name: "index___en"
  },
  {
    path: "/fr/",
    component: _3237362a,
    name: "index___fr"
  },
  {
    path: "/about",
    component: _71a6ebb4,
    name: "about___en"
  },
  {
    path: "/fr/about",
    component: _71a6ebb4,
    name: "about___fr"
  }
]
```

Note that routes for the English version do not have any prefix because it is the default language, see [strategies](/strategies) for more details.
