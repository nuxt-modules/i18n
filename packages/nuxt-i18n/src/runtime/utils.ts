import { isVue3 } from 'vue-demi'
import { useRequestHeaders } from '#app'
import { findBrowserLocale } from 'vue-i18n-routing'

import type { Composer } from '@intlify/vue-i18n-bridge'
import type { NuxtI18nInternalOptions } from '#build/i18n.options.mjs'

/**
 * Parses locales provided from browser through `accept-language` header.
 *
 * @param input - Accept-Language header value.
 * @return An array of locale codes. Priority determined by order in array.
 */
export function parseAcceptLanguage(input: string): string[] {
  // Example input: en-US,en;q=0.9,nb;q=0.8,no;q=0.7
  // Contains tags separated by comma.
  // Each tag consists of locale code (2-3 letter language code) and optionally country code
  // after dash. Tag can also contain score after semicolon, that is assumed to match order
  // so it's not explicitly used.
  return input.split(',').map(tag => tag.split(';')[0])
}

export async function loadAndSetLocale(newLocale: string, i18n: Composer /*, { initialSetup = false } = {}*/) {
  if (!newLocale) {
    return
  }

  // abort if different domains option enabled
  // if (!initialSetup && i18n.differentDomains) {
  //   return
  // }

  const oldLocale = i18n.locale.value as string
  if (newLocale === oldLocale) {
    return
  }

  i18n.locale.value = newLocale
  console.log('loadAndSetLocale', newLocale, oldLocale, i18n)
}

export function getBrowserLocale(options: Required<NuxtI18nInternalOptions>, context?: any): string | undefined {
  let ret: string | undefined

  if (process.client) {
    if (navigator.languages) {
      // get browser language either from navigator if running on client side, or from the headers
      ret = findBrowserLocale(options.__normalizedLocales, navigator.languages as string[])
    }
  } else if (process.server) {
    if (!isVue3) {
      if (context.req && typeof context.req.headers['accept-language'] !== 'undefined') {
        ret = findBrowserLocale(
          options.__normalizedLocales,
          parseAcceptLanguage(context.req.headers['accept-language'])
        )
      }
    } else {
      const headers = useRequestHeaders(['accept-language'])
      if (headers['accept-language']) {
        ret = findBrowserLocale(options.__normalizedLocales, parseAcceptLanguage(headers['accept-language']))
      }
    }
  }

  return ret
}
