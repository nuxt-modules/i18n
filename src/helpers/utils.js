/**
 * @typedef {import('../../types/internal').ResolvedOptions} ResolvedOptions
 */

/**
 * Retrieve page's options from the module's configuration for a given route
 *
 * @typedef {{ locales: readonly string[], paths: Record<string, string>} } ComputedPageOptions
 *
 * @param {import('@nuxt/types/config/router').NuxtRouteConfig} route
 * @param {ResolvedOptions['pages']} pages Pages options from module's configuration
 * @param {ResolvedOptions['localeCodes']} localeCodes
 * @param {string} pagesDir Pages dir from Nuxt's configuration
 * @param {ResolvedOptions['defaultLocale']} defaultLocale Default locale from Nuxt's configuration
 * @return {ComputedPageOptions | false} Page options
 */
export function getPageOptions (route, pages, localeCodes, pagesDir, defaultLocale) {
  /** @type {ComputedPageOptions} */
  const options = {
    locales: localeCodes,
    paths: {}
  }
  const pattern = new RegExp(`${pagesDir}/`, 'i')
  const chunkName = route.chunkName ? route.chunkName.replace(pattern, '') : route.name
  const pageOptions = chunkName ? pages[chunkName] : undefined
  // Routing disabled
  if (pageOptions === false) {
    return false
  }
  // Skip if no page options defined
  if (!pageOptions) {
    return options
  }

  // Remove disabled locales from page options
  options.locales = options.locales.filter(locale => pageOptions[locale] !== false)

  // Construct paths object
  for (const locale of options.locales) {
    const customLocalePath = pageOptions[locale]
    if (typeof customLocalePath === 'string') {
      // Set custom path if any
      options.paths[locale] = customLocalePath
      continue
    }

    const customDefaultLocalePath = pageOptions[defaultLocale]
    if (typeof customDefaultLocalePath === 'string') {
      // Set default locale's custom path if any
      options.paths[locale] = customDefaultLocalePath
    }
  }

  return options
}

/**
 * @param {string} routePath
 * @param {boolean | undefined} trailingSlash
 * @param {boolean} [isChildWithRelativePath] Indicates if it is a child route that has relative path
 * @return {string}
 */
export function adjustRouteDefinitionForTrailingSlash (routePath, trailingSlash, isChildWithRelativePath) {
  return routePath.replace(/\/+$/, '') + (trailingSlash ? '/' : '') || (isChildWithRelativePath ? '' : '/')
}
