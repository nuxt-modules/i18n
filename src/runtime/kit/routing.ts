import type { BrowserLocale, Locale, RouteName, RouteObject } from './types'

/**
 * The browser locale matcher
 *
 * @param locales - The target {@link LocaleObject | locale} list
 * @param browserLocales - The locale code list that is used in browser
 *
 * @returns The matched {@link BrowserLocale | locale info}
 */
function matchBrowserLocale(locales: Locale[], browserLocales: readonly string[]): BrowserLocale[] {
  const matchedLocales: BrowserLocale[] = []

  // first pass: match exact locale.
  for (const [index, browserCode] of browserLocales.entries()) {
    const matchedLocale = locales.find(l => l.language?.toLowerCase() === browserCode.toLowerCase())
    if (matchedLocale) {
      matchedLocales.push({ code: matchedLocale.code, score: 1 - index / browserLocales.length })
      break
    }
  }

  // second pass: match only locale code part of the browser locale (not including country).
  for (const [index, browserCode] of browserLocales.entries()) {
    const languageCode = browserCode.split('-')[0].toLowerCase()
    const matchedLocale = locales.find(l => l.language?.split('-')[0].toLowerCase() === languageCode)
    if (matchedLocale) {
      // deduct a thousandth for being non-exact match.
      matchedLocales.push({ code: matchedLocale.code, score: 0.999 - index / browserLocales.length })
      break
    }
  }

  return matchedLocales
}

function compareBrowserLocale(a: BrowserLocale, b: BrowserLocale): number {
  if (a.score === b.score) {
    // if scores are equal then pick more specific (longer) code.
    return b.code.length - a.code.length
  }
  return b.score - a.score
}

/**
 * Find the browser locale
 *
 * @param locales - The target {@link Locale} list
 * @param browserLocales - The locale code list that is used in browser
 *
 * @returns The matched the locale code
 * @internal
 */
export function findBrowserLocale(locales: Locale[], browserLocales: readonly string[]): string {
  const normalizedLocales = locales.map(l => ({ code: l.code, language: l.language || l.code }))
  const matchedLocales = matchBrowserLocale(normalizedLocales, browserLocales)
  if (matchedLocales.length === 0) {
    return ''
  }

  // sort by score when multiple locales matched
  if (matchedLocales.length > 1) {
    matchedLocales.sort(compareBrowserLocale)
  }

  return matchedLocales[0].code
}

export function getRouteNameLocaleRegex(
  options: { localeCodes: string[]; separator: string; defaultSuffix: string },
  localesPattern: string = getLocalesPattern(options.localeCodes)
) {
  const defaultSuffixPattern = `(?:${options.separator}${options.defaultSuffix})?`
  return new RegExp(`${options.separator}${localesPattern}${defaultSuffixPattern}$`, 'i')
}

/**
 * Convenience function to return the first match of the regex or empty string
 * @internal
 */
export function createNameLocaleRegexMatcher(re: RegExp) {
  return (val: string) => val.match(re)?.[1] ?? ''
}

/**
 * Normalizes {@link RouteName} to string
 * @internal
 */
export function normalizeRouteName(routeName: RouteName) {
  if (typeof routeName === 'string') return routeName
  if (routeName != null) return routeName.toString()
  return ''
}

/**
 * Normalizes {@link RouteName} or {@link RouteObject} to string
 * @internal
 */
export function getRouteName(route: RouteName | RouteObject) {
  if (typeof route === 'object') return normalizeRouteName(route?.name)
  return normalizeRouteName(route)
}

/**
 * Extract route name without localization from {@link RouteName} or {@link RouteObject}
 * @internal
 */
export function getRouteBaseName(route: RouteName | RouteObject, separator: string) {
  return getRouteName(route).split(separator)[0]
}

export function getLocalizedRouteName(
  routeName: string,
  locale: string,
  isDefault: boolean,
  separator: string = '___',
  defaultSuffix: string = 'default'
) {
  if (isDefault) {
    return routeName + separator + locale + separator + defaultSuffix
  }
  return routeName + separator + locale
}

/**
 * Match locale code from route path (e.g. `/en/about` => `en`)
 * @internal
 */
export function getRoutePathLocaleRegex(
  localeCodes: string[],
  localesPattern: string = getLocalesPattern(localeCodes)
) {
  return new RegExp(`^/${localesPattern}(?:/|$)`, 'i')
}

function getLocalesPattern(localeCodes: string[]) {
  return `(${localeCodes.join('|')})`
}

/**
 * NOTE: this likely needs to be implemented on the utility function consumer side
 * @internal
 */
export function createLocaleFromRouteGetter(options: {
  localeCodes: string[]
  separator: string
  defaultSuffix: string
}) {
  const localesPattern = getLocalesPattern(options.localeCodes)
  const regexpName = getRouteNameLocaleRegex(options, localesPattern)
  const regexpPath = getRoutePathLocaleRegex(options.localeCodes, localesPattern)

  const matchPath = createNameLocaleRegexMatcher(regexpPath)
  const matchName = createNameLocaleRegexMatcher(regexpName)

  /**
   * extract locale code from route name or path
   */
  return (route: RouteName | RouteObject) => {
    if (typeof route === 'string') {
      return matchPath(route)
    }

    if (typeof route === 'symbol') {
      return matchPath(String(route))
    }

    // extract from route name
    if (route?.name) {
      return matchName(normalizeRouteName(route.name))
    }

    // extract from path
    if (route?.path) {
      return matchPath(route.path)
    }

    return ''
  }
}
