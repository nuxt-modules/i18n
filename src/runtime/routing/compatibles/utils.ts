import { useRouter } from '#imports'
import { assign } from '@intlify/shared'

import type { Locale } from 'vue-i18n'
import type { RouteLocationNormalizedLoaded, RouteLocationPathRaw } from 'vue-router'
import type { Strategies } from '#build/i18n.options.mjs'

function split(str: string, index: number) {
  const result = [str.slice(0, index), str.slice(index)]
  return result
}

/**
 * NOTE:
 * Nuxt route uses a proxy with getters for performance reasons (https://github.com/nuxt/nuxt/pull/21957).
 * Spreading will result in an empty object, so we make a copy of the route by accessing each getter property by name.
 */
export function routeToObject(route: RouteLocationNormalizedLoaded) {
  const { fullPath, query, hash, name, path, params, meta, redirectedFrom, matched } = route
  return {
    fullPath,
    params,
    query,
    hash,
    name,
    path,
    meta,
    matched,
    redirectedFrom
  }
}

/**
 * NOTE:
 * vue-router v4.x `router.resolve` for a non exists path will output a warning.
 * `router.hasRoute`, which checks for the route can only be a named route.
 * When using the `prefix` strategy, the path specified by `localePath` is specified as a path not prefixed with a locale.
 * This will cause vue-router to issue a warning, so we can work-around by using `router.options.routes`.
 */
export function resolve(route: RouteLocationPathRaw, strategy: Strategies, locale: Locale) {
  const router = useRouter()
  if (strategy !== 'prefix') {
    return router.resolve(route)
  }

  // if (isArray(route.matched) && route.matched.length > 0) {
  //   return route.matched[0]
  // }

  const [rootSlash, restPath] = split(route.path, 1)
  const targetPath = `${rootSlash}${locale}${restPath === '' ? restPath : `/${restPath}`}`
  const _route = router.options?.routes?.find(r => r.path === targetPath)

  if (_route == null) {
    return route
  }

  const _resolvableRoute = assign({}, route, _route)
  _resolvableRoute.path = targetPath
  return router.resolve(_resolvableRoute)
}
