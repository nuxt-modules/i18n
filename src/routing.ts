import { getNormalizedLocales } from './utils'
import { isObject } from '@intlify/shared'

import type { Locale } from 'vue-i18n'
import type { NuxtPage } from '@nuxt/schema'
import type { MarkRequired, MarkOptional } from 'ts-essentials'
import type { NuxtI18nOptions, PrefixLocalizedRouteOptions, RouteOptionsResolver } from './types'

const join = (...args: (string | undefined)[]) => args.filter(Boolean).join('')

/**
 * Options to compute route localizing
 *
 * @remarks
 * The route options that is compute the route to be localized on {@link localizeRoutes}
 *
 * @public
 */
export declare interface ComputedRouteOptions {
  locales: readonly string[]
  paths: Record<string, string>
}

export function prefixLocalizedRoute(
  localizeOptions: PrefixLocalizedRouteOptions,
  options: LocalizeRoutesParams,
  extra = false
): boolean {
  const isDefaultLocale = localizeOptions.locale === (localizeOptions.defaultLocale ?? '')
  const isChildWithRelativePath = localizeOptions.parent != null && !localizeOptions.path.startsWith('/')

  // no need to add prefix if child's path is relative
  return (
    !extra &&
    !isChildWithRelativePath &&
    // skip default locale if strategy is 'prefix_except_default'
    !(isDefaultLocale && options.strategy === 'prefix_except_default')
  )
}

function adjustRoutePathForTrailingSlash(localized: LocalizedRoute, trailingSlash: boolean) {
  const isChildWithRelativePath = localized.parent != null && !localized.path.startsWith('/')
  return localized.path.replace(/\/+$/, '') + (trailingSlash ? '/' : '') || (isChildWithRelativePath ? '' : '/')
}

export type LocalizeRoutesParams = MarkRequired<
  NuxtI18nOptions,
  'strategy' | 'locales' | 'routesNameSeparator' | 'trailingSlash' | 'defaultLocaleRouteNameSuffix'
> & {
  includeUnprefixedFallback?: boolean
  optionsResolver?: RouteOptionsResolver
}

type LocalizedRoute = NuxtPage & { locale: Locale; parent: NuxtPage | undefined }
type LocalizeRouteParams = {
  /**
   * locales to use for localization
   */
  locales: string[]
  /**
   * parent route
   */
  parent?: NuxtPage
  /**
   * localized parent route
   */
  parentLocalized?: NuxtPage
  /**
   * indicates whether this is a default route for 'prefix_and_default' strategy
   */
  extra?: boolean
}

/**
 * Localize routes
 *
 * @param routes - Some routes
 * @param options - An options
 *
 * @returns Localized routes
 *
 * @public
 */
export function localizeRoutes(routes: NuxtPage[], options: LocalizeRoutesParams): NuxtPage[] {
  if (options.strategy === 'no_prefix') {
    return routes
  }

  let defaultLocales = [options.defaultLocale ?? '']
  if (options.differentDomains) {
    const domainDefaults = options.locales
      .filter(locale => (isObject(locale) ? locale.domainDefault : false))
      .map(locale => (isObject(locale) ? locale.code : locale))
    defaultLocales = defaultLocales.concat(domainDefaults)
  }

  function localizeRoute(
    route: NuxtPage,
    { locales = [], parent, parentLocalized, extra = false }: LocalizeRouteParams
  ): NuxtPage[] {
    // skip route localization
    if (route.redirect && !route.file) {
      return [route]
    }

    // resolve with route (page) options
    const routeOptions = options.optionsResolver?.(route, locales)
    if (options.optionsResolver != null && routeOptions == null) {
      return [route]
    }

    // component specific options
    const componentOptions: ComputedRouteOptions = {
      // filter locales to prevent child routes from being localized even though they are disabled in the configuration.
      locales: locales.filter(locale => (routeOptions?.locales ?? locales).includes(locale)),
      paths: {},
      ...routeOptions
    }

    const localizedRoutes: (LocalizedRoute | NuxtPage)[] = []
    for (const locale of componentOptions.locales) {
      const localized: LocalizedRoute = { ...route, locale, parent }
      const isDefaultLocale = defaultLocales.includes(locale)
      const addDefaultTree = isDefaultLocale && options.strategy === 'prefix_and_default' && parent == null && !extra

      // localize route again for strategy `prefix_and_default`
      if (addDefaultTree && parent == null && !extra) {
        localizedRoutes.push(...localizeRoute(route, { locales: [locale], extra: true }))
      }

      const nameSegments = [localized.name, options.routesNameSeparator, locale]
      if (extra) {
        nameSegments.push(options.routesNameSeparator, options.defaultLocaleRouteNameSuffix)
      }

      // localize name if set
      localized.name &&= join(...nameSegments)

      // use custom path if found
      localized.path = componentOptions.paths?.[locale] ?? localized.path

      const localePrefixable = prefixLocalizedRoute(
        { defaultLocale: isDefaultLocale ? locale : options.defaultLocale, ...localized },
        options,
        extra
      )
      if (localePrefixable) {
        localized.path = join('/', locale, localized.path)

        if (isDefaultLocale && options.strategy === 'prefix' && options.includeUnprefixedFallback) {
          localizedRoutes.push({ ...route, locale, parent })
        }
      }

      localized.path &&= adjustRoutePathForTrailingSlash(localized, options.trailingSlash)

      // remove parent path from child route
      if (parentLocalized != null) {
        localized.path = localized.path.replace(parentLocalized.path + '/', '')
      }

      // localize child routes if set
      localized.children &&= localized.children.flatMap(child =>
        localizeRoute(child, { locales: [locale], parent: route, parentLocalized: localized, extra })
      )

      localizedRoutes.push(localized)
    }

    // remove properties used for localization process
    return localizedRoutes.flatMap((x: MarkOptional<LocalizedRoute, 'parent' | 'locale'>) => {
      delete x.parent
      delete x.locale
      return x
    })
  }

  return routes.flatMap(route =>
    localizeRoute(route, { locales: getNormalizedLocales(options.locales).map(x => x.code) })
  )
}
