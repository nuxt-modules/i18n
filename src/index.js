import { resolve, join } from 'path'
import { readdirSync } from 'fs'
import merge from 'lodash.merge'
// @ts-ignore
import { directive as i18nExtensionsDirective } from '@intlify/vue-i18n-extensions'
import { COMPONENT_OPTIONS_KEY, DEFAULT_OPTIONS, ROOT_DIR, STRATEGIES, REDIRECT_ON_OPTIONS } from './helpers/constants'
import { buildHook, createExtendRoutesHook } from './core/hooks'
import { formatMessage } from './templates/utils-common'

/** @type {import('@nuxt/types').Module<import('../types').Options>} */
export default async function (moduleOptions) {
  /** @type {import('../types/internal').ResolvedOptions} */
  const options = merge({}, DEFAULT_OPTIONS, moduleOptions, this.options.i18n)

  if (!Object.values(STRATEGIES).includes(options.strategy)) {
    // eslint-disable-next-line no-console
    console.error(formatMessage(`Invalid "strategy" option "${options.strategy}" (must be one of: ${Object.values(STRATEGIES).join(', ')}).`))
    return
  }

  if (options.lazy && !options.langDir) {
    throw new Error(formatMessage('When using the "lazy" option you must also set the "langDir" option.'))
  }

  if (options.langDir) {
    if (!options.locales.length || typeof options.locales[0] === 'string') {
      throw new Error(formatMessage('When using the "langDir" option the "locales" must be a list of objects.'))
    }
    for (const locale of options.locales) {
      if (typeof (locale) === 'string' || !locale.file) {
        throw new Error(formatMessage(`All locales must be objects and have the "file" property set when using "langDir".\nFound none in:\n${JSON.stringify(locale, null, 2)}.`))
      }
    }
    options.langDir = this.nuxt.resolver.resolveAlias(options.langDir)
  }

  // Templates (including plugins).
  // This is done here rather than in the build hook to ensure the order the plugins are added
  // is predictable between different modules.
  const nuxtOptions = this.options
  const normalizedLocales = []
  for (const locale of options.locales) {
    if (typeof (locale) === 'string') {
      normalizedLocales.push({ code: locale })
    } else {
      normalizedLocales.push(locale)
    }
  }
  options.normalizedLocales = normalizedLocales
  // Get an array of locale codes from the list of locales.
  options.localeCodes = options.normalizedLocales.map(locale => locale.code)

  const templatesOptions = {
    Constants: {
      COMPONENT_OPTIONS_KEY,
      STRATEGIES,
      REDIRECT_ON_OPTIONS
    },
    nuxtOptions: {
      isUniversalMode: nuxtOptions.mode === 'universal',
      trailingSlash: nuxtOptions.router.trailingSlash
    },
    options
  }

  const templatesPath = join(__dirname, '/templates')
  for (const file of readdirSync(templatesPath)) {
    if (!file.endsWith('.js')) {
      continue
    }
    if (file.startsWith('plugin.')) {
      this.addPlugin({
        src: resolve(templatesPath, file),
        fileName: join(ROOT_DIR, file),
        options: templatesOptions
      })
    } else {
      this.addTemplate({
        src: resolve(templatesPath, file),
        fileName: join(ROOT_DIR, file),
        options: templatesOptions
      })
    }
  }

  if (options.strategy !== STRATEGIES.NO_PREFIX && options.localeCodes.length) {
    this.extendRoutes(createExtendRoutesHook.call(this, options))
  }

  await this.nuxt.hook('build:before', () => buildHook.call(this, options))

  this.options.alias['~i18n-klona'] = require.resolve('klona/full').replace(/\.js$/, '.mjs')
  this.options.alias['~i18n-ufo'] = require.resolve('ufo').replace(/\.js$/, '.mjs')

  if (!Array.isArray(this.options.router.middleware)) {
    throw new TypeError(formatMessage('options.router.middleware is not an array.'))
  }
  this.options.router.middleware.push('nuxti18n')

  if (!this.options.render.bundleRenderer || typeof (this.options.render.bundleRenderer) !== 'object') {
    throw new TypeError(formatMessage('options.render.bundleRenderer is not an object.'))
  }
  this.options.render.bundleRenderer.directives = this.options.render.bundleRenderer.directives || {}
  this.options.render.bundleRenderer.directives.t = i18nExtensionsDirective

  // Transpile is-https (IE11)
  this.options.build.transpile = this.options.build.transpile || []
  this.options.build.transpile.push('is-https')
}
