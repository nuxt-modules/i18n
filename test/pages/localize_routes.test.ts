import { describe, it, expect } from 'vitest'

import { localizeRoutes } from '../../src/routing'
import { getNuxtOptions } from './utils'

import type { NuxtPage } from '@nuxt/schema'
import type { LocaleObject } from 'vue-i18n-routing'

const nuxtOptions = getNuxtOptions({})
nuxtOptions.locales = nuxtOptions.locales?.filter(x => (x as LocaleObject).code !== 'fr') as LocaleObject[]
delete nuxtOptions.defaultLocale

describe('localizeRoutes', function () {
  describe('basic', function () {
    it('should be localized routing', function () {
      const routes: NuxtPage[] = [
        {
          path: '/',
          name: 'home'
        },
        {
          path: '/about',
          name: 'about'
        }
      ]
      const localeCodes = ['en', 'ja']
      const localizedRoutes = localizeRoutes(routes, { ...nuxtOptions, locales: localeCodes })

      expect(localizedRoutes).toMatchSnapshot()
      expect(localizedRoutes.length).to.equal(4)
      localeCodes.forEach(locale => {
        routes.forEach(route => {
          expect(localizedRoutes).to.deep.include({
            path: `/${locale}${route.path === '/' ? '' : route.path}`,
            name: `${route.name}${nuxtOptions.routesNameSeparator}${locale}`
          })
        })
      })
    })
  })

  describe('has children', function () {
    it('should be localized routing', function () {
      const routes: NuxtPage[] = [
        {
          path: '/user/:id',
          name: 'user',
          children: [
            {
              path: 'profile',
              name: 'user-profile'
            },
            {
              path: 'posts',
              name: 'user-posts'
            }
          ]
        }
      ]
      const children: NuxtPage[] = routes[0].children as NuxtPage[]

      const localeCodes = ['en', 'ja']
      const localizedRoutes = localizeRoutes(routes, { ...nuxtOptions, locales: localeCodes })

      expect(localizedRoutes).toMatchSnapshot()
      expect(localizedRoutes.length).to.equal(2)
      localeCodes.forEach(locale => {
        routes.forEach(route => {
          expect(localizedRoutes).to.deep.include({
            path: `/${locale}${route.path === '/' ? '' : route.path}`,
            name: `${route.name}${nuxtOptions.routesNameSeparator}${locale}`,
            children: children.map(child => ({
              path: child.path,
              name: `${child.name}${nuxtOptions.routesNameSeparator}${locale}`
            }))
          })
        })
      })
    })
  })

  describe('trailing slash', function () {
    it('should be localized routing', function () {
      const routes: NuxtPage[] = [
        {
          path: '/',
          name: 'home'
        },
        {
          path: '/about',
          name: 'about'
        }
      ]
      const localeCodes = ['en', 'ja']
      const localizedRoutes = localizeRoutes(routes, { ...nuxtOptions, locales: localeCodes, trailingSlash: true })

      expect(localizedRoutes).toMatchSnapshot()
      expect(localizedRoutes.length).to.equal(4)
      localeCodes.forEach(locale => {
        routes.forEach(route => {
          expect(localizedRoutes).to.deep.include({
            path: `/${locale}${route.path === '/' ? '' : route.path}/`,
            name: `${route.name}${nuxtOptions.routesNameSeparator}${locale}`
          })
        })
      })
    })
  })

  describe('route name separator', function () {
    it('should be localized routing', function () {
      const routes: NuxtPage[] = [
        {
          path: '/',
          name: 'home'
        },
        {
          path: '/about',
          name: 'about'
        }
      ]
      const localeCodes = ['en', 'ja']
      const localizedRoutes = localizeRoutes(routes, {
        ...nuxtOptions,
        locales: localeCodes,
        routesNameSeparator: '__'
      })

      expect(localizedRoutes).toMatchSnapshot()
      expect(localizedRoutes.length).to.equal(4)
      localeCodes.forEach(locale => {
        routes.forEach(route => {
          expect(localizedRoutes).to.deep.include({
            path: `/${locale}${route.path === '/' ? '' : route.path}`,
            name: `${route.name}${'__'}${locale}`
          })
        })
      })
    })
  })

  describe('strategy: "prefix_and_default"', function () {
    it('should be localized routing', function () {
      const routes: NuxtPage[] = [
        {
          path: '/',
          name: 'home'
        },
        {
          path: '/about',
          name: 'about'
        },
        {
          path: '/user/:id',
          name: 'user',
          children: [
            {
              path: 'profile',
              name: 'user-profile'
            },
            {
              path: 'posts',
              name: 'user-posts'
            }
          ]
        }
      ]

      const localeCodes = ['en', 'ja']
      const localizedRoutes = localizeRoutes(routes, {
        ...nuxtOptions,
        defaultLocale: 'en',
        strategy: 'prefix_and_default',
        locales: localeCodes
      })

      expect(localizedRoutes).toMatchSnapshot()
    })
  })

  describe('strategy: "prefix_except_default"', function () {
    it('should be localized routing', function () {
      const routes: NuxtPage[] = [
        {
          path: '/',
          name: 'home'
        },
        {
          path: '/about',
          name: 'about'
        }
      ]
      const localeCodes = ['en', 'ja']
      const localizedRoutes = localizeRoutes(routes, {
        ...nuxtOptions,
        defaultLocale: 'en',
        strategy: 'prefix_except_default',
        locales: localeCodes
      })

      expect(localizedRoutes).toMatchSnapshot()
    })
  })

  describe('strategy: "prefix"', function () {
    it('should be localized routing', function () {
      const routes: NuxtPage[] = [
        {
          path: '/',
          name: 'home'
        },
        {
          path: '/about',
          name: 'about'
        }
      ]
      const localeCodes = ['en', 'ja']
      const localizedRoutes = localizeRoutes(routes, {
        ...nuxtOptions,
        defaultLocale: 'en',
        strategy: 'prefix',
        locales: localeCodes,
        includeUnprefixedFallback: true
      })

      expect(localizedRoutes).toMatchSnapshot()
    })
  })

  describe('strategy: "no_prefix"', function () {
    it('should be localized routing', function () {
      const routes: NuxtPage[] = [
        {
          path: '/',
          name: 'home'
        },
        {
          path: '/about',
          name: 'about'
        }
      ]
      const localeCodes = ['en', 'ja']
      const localizedRoutes = localizeRoutes(routes, {
        ...nuxtOptions,
        defaultLocale: 'en',
        strategy: 'no_prefix',
        locales: localeCodes
      })

      expect(localizedRoutes).toMatchSnapshot()
    })
  })

  describe('Route options resolver: routing disable', () => {
    it('should be disabled routing', () => {
      const routes: NuxtPage[] = [
        {
          path: '/',
          name: 'home'
        },
        {
          path: '/about',
          name: 'about'
        }
      ]
      const localeCodes = ['en', 'ja']
      const localizedRoutes = localizeRoutes(routes, {
        ...nuxtOptions,
        locales: localeCodes,
        optionsResolver: () => undefined
      })

      expect(localizedRoutes).toMatchSnapshot()
    })
  })
})
