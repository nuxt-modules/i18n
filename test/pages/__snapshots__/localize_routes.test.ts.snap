// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`localizeRoutes > Route options resolver: routing disable > should be disabled routing 1`] = `
[
  {
    "name": "home",
    "path": "/",
  },
  {
    "name": "about",
    "path": "/about",
  },
]
`;

exports[`localizeRoutes > basic > should be localized routing 1`] = `
[
  {
    "name": "home___en",
    "path": "/en",
  },
  {
    "name": "home___ja",
    "path": "/ja",
  },
  {
    "name": "about___en",
    "path": "/en/about",
  },
  {
    "name": "about___ja",
    "path": "/ja/about",
  },
]
`;

exports[`localizeRoutes > has children > should be localized routing 1`] = `
[
  {
    "children": [
      {
        "name": "user-profile___en",
        "path": "profile",
      },
      {
        "name": "user-posts___en",
        "path": "posts",
      },
    ],
    "name": "user___en",
    "path": "/en/user/:id",
  },
  {
    "children": [
      {
        "name": "user-profile___ja",
        "path": "profile",
      },
      {
        "name": "user-posts___ja",
        "path": "posts",
      },
    ],
    "name": "user___ja",
    "path": "/ja/user/:id",
  },
]
`;

exports[`localizeRoutes > route name separator > should be localized routing 1`] = `
[
  {
    "name": "home__en",
    "path": "/en",
  },
  {
    "name": "home__ja",
    "path": "/ja",
  },
  {
    "name": "about__en",
    "path": "/en/about",
  },
  {
    "name": "about__ja",
    "path": "/ja/about",
  },
]
`;

exports[`localizeRoutes > strategy: "no_prefix" > should be localized routing 1`] = `
[
  {
    "name": "home",
    "path": "/",
  },
  {
    "name": "about",
    "path": "/about",
  },
]
`;

exports[`localizeRoutes > strategy: "prefix" > should be localized routing 1`] = `
[
  {
    "name": "home",
    "path": "/",
  },
  {
    "name": "home___en",
    "path": "/en",
  },
  {
    "name": "home___ja",
    "path": "/ja",
  },
  {
    "name": "about",
    "path": "/about",
  },
  {
    "name": "about___en",
    "path": "/en/about",
  },
  {
    "name": "about___ja",
    "path": "/ja/about",
  },
]
`;

exports[`localizeRoutes > strategy: "prefix_and_default" > should be localized routing 1`] = `
[
  {
    "name": "home___en___default",
    "path": "/",
  },
  {
    "name": "home___en",
    "path": "/en",
  },
  {
    "name": "home___ja",
    "path": "/ja",
  },
  {
    "name": "about___en___default",
    "path": "/about",
  },
  {
    "name": "about___en",
    "path": "/en/about",
  },
  {
    "name": "about___ja",
    "path": "/ja/about",
  },
  {
    "children": [
      {
        "name": "user-profile___en___default",
        "path": "profile",
      },
      {
        "name": "user-posts___en___default",
        "path": "posts",
      },
    ],
    "name": "user___en___default",
    "path": "/user/:id",
  },
  {
    "children": [
      {
        "name": "user-profile___en",
        "path": "profile",
      },
      {
        "name": "user-posts___en",
        "path": "posts",
      },
    ],
    "name": "user___en",
    "path": "/en/user/:id",
  },
  {
    "children": [
      {
        "name": "user-profile___ja",
        "path": "profile",
      },
      {
        "name": "user-posts___ja",
        "path": "posts",
      },
    ],
    "name": "user___ja",
    "path": "/ja/user/:id",
  },
]
`;

exports[`localizeRoutes > strategy: "prefix_except_default" > should be localized routing 1`] = `
[
  {
    "name": "home___en",
    "path": "/",
  },
  {
    "name": "home___ja",
    "path": "/ja",
  },
  {
    "name": "about___en",
    "path": "/about",
  },
  {
    "name": "about___ja",
    "path": "/ja/about",
  },
]
`;

exports[`localizeRoutes > trailing slash > should be localized routing 1`] = `
[
  {
    "name": "home___en",
    "path": "/en/",
  },
  {
    "name": "home___ja",
    "path": "/ja/",
  },
  {
    "name": "about___en",
    "path": "/en/about/",
  },
  {
    "name": "about___ja",
    "path": "/ja/about/",
  },
]
`;
